src/main/java/frc/robot/subsystems/TurretSubsystem.java

```diff
@@ -0,0 +1,246 @@
+import edu.wpi.first.wpilibj2.command.InstantCommand;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants;
+import frc.robot.Robot;
+import frc.robot.stateManagement.AllianceColor;
+import frc.robot.stateManagement.RobotStateManager;
+import frc.robot.utilities.DebugEntry;
+import frc.robot.utilities.LimelightHelpers;
+import java.util.function.Consumer;
+
+public class TurretSubsystem extends SubsystemBase {
+
+  private CANSparkMax turretMotor;
+  private SingleJointedArmSim turretSim;
+  private Mechanism2d turretMech;
+  private MechanismRoot2d root;
+  private MechanismLigament2d turretAngleSim;
+  private ShuffleboardTab turretTab = Shuffleboard.getTab("Turret Tab");
+  private SimDeviceSim simEncoder;
+  private SimDouble simTurretPos;
+
+  private PIDController turretPIDController;
+  private CANcoder m_encoder;
+  private double turretPosition;
+  private double turretVelocity;
+  private Consumer<Double> testPosition;
+  private ShuffleboardTab configTab = Shuffleboard.getTab("Config");
+  private GenericEntry kP = configTab.add("Turret KP", Constants.TurretConstants.KP).getEntry();
+  private GenericEntry kI = configTab.add("Turret KI", Constants.TurretConstants.KI).getEntry();
+  private GenericEntry kD = configTab.add("Turret KD", Constants.TurretConstants.KD).getEntry();
+  private DebugEntry<Double> turretPositionEntry =
+      new DebugEntry<Double>(turretPosition, "Position", this);
+  private DebugEntry<Double> turretGoalPositionEntry =
+      new DebugEntry<Double>(0.0, "Goal Position", this);
+  private DebugEntry<Double> turretVelocityEntry =
+      new DebugEntry<Double>(turretVelocity, "Velocity", this);
+
+  private DebugEntry<Double> tagDistanceEntry =
+      new DebugEntry<Double>(0.0, "LastMeasuredTagDistance", this);
+
+  private final RobotStateManager robotStateManager;
+
+  public TurretSubsystem(RobotStateManager robotStateManager) {
+    turretMotor = new CANSparkMax(Constants.TurretConstants.MOTOR_ID, MotorType.kBrushless);
+
+    // Simulation
+    if (Robot.isSimulation()) {
+      turretSim =
+          new SingleJointedArmSim(
+              DCMotor.getNEO(1),
+              4,
+              3.5 * 0.1016 * 0.1016 / 3,
+              0.1016,
+              -Math.toRadians(Constants.TurretConstants.MAX_TURRET_ANGLE_DEGREES),
+              Math.toRadians(Constants.TurretConstants.MAX_TURRET_ANGLE_DEGREES),
+              false,
+              0);
+    }
+    turretMech = new Mechanism2d(4, 4);
+    root = turretMech.getRoot("Root", 2, 2);
+    turretAngleSim =
+        root.append(new MechanismLigament2d("Turret", 2, 0, 5, new Color8Bit(Color.kRed)));
+    turretTab.add("Turret", turretMech);
+
+    turretMotor.restoreFactoryDefaults();
+    turretMotor.setSmartCurrentLimit(40);
+
+    this.robotStateManager = robotStateManager;
+
+    turretMotor.setSoftLimit(CANSparkMax.SoftLimitDirection.kReverse, -2);
+    turretMotor.setSoftLimit(CANSparkMax.SoftLimitDirection.kForward, 2);
```

Comment:
The numbers here should be related to our max angle and conversion factor. 

Also verify that using negative numbers for kReverse is correct.